src/main/java/edu/kit/provideq/toolbox/api/SolveRouter.java

```diff
@@ -0,0 +1,143 @@
+                .content(contentBuilder()
+                    .schema(schemaBuilder().implementation(
+                        metaSolver.getProblemType().getRequestType()))
+                    .mediaType(APPLICATION_JSON_VALUE)
+                )
+                .required(true)
+            )
+            .response(responseBuilder()
+                .responseCode(String.valueOf(HttpStatus.OK.value()))
+                .implementation(SolutionHandle.class)
+            )
+    ).build();
+  }
+
+  private <ProblemT, SolutionT> Mono<ServerResponse> handleRouteForMetaSolver(
+      MetaSolver<ProblemT, SolutionT, ?> metaSolver, ServerRequest req) {
+    var solutionMono = req
+        .bodyToMono(new ParameterizedTypeReference<SolveRequest<ProblemT>>() {
+        })
+        .doOnNext(this::validate)
+        .map(metaSolver::solve)
+        .map(Solution::toStringSolution);
+    return ok().body(solutionMono, new ParameterizedTypeReference<>() {
+    });
+  }
+
+  private <ProblemT> void validate(SolveRequest<ProblemT> request) {
+    Errors errors = new BeanPropertyBindingResult(request, "request");
+    validator.validate(request, errors);
+    if (errors.hasErrors()) {
+      throw new ServerWebInputException(errors.toString());
+    }
+  }
+
+  @Bean
+  RouterFunction<ServerResponse> getSolutionRoutes() {
+    return metaSolverProvider.getMetaSolvers().stream()
+        .map(this::defineSolutionRouteForMetaSolver)
+        .reduce(RouterFunction::and)
+        .orElseThrow(); // we should always have at least one route or the toolbox is useless
+  }
+
+  private RouterFunction<ServerResponse> defineSolutionRouteForMetaSolver(
+      MetaSolver<?, ?, ?> metaSolver) {
+    String problemId = metaSolver.getProblemType().getId();
+    return route().GET(
+        "/solve/" + problemId,
+        accept(APPLICATION_JSON),
+        req -> handleSolutionRouteForMetaSolver(metaSolver, req),
+        ops -> ops
+            .operationId("/solution/" + problemId)
+            .tag(problemId)
+            .parameter(parameterBuilder().in(ParameterIn.QUERY).name("id"))
+            .response(responseBuilder()
+                .responseCode(String.valueOf(HttpStatus.OK.value()))
+                .implementation(SolutionHandle.class)
+            )
+    ).build();
+  }
+
+  private Mono<ServerResponse> handleSolutionRouteForMetaSolver(MetaSolver<?, ?, ?> metaSolver,
+                                                                ServerRequest req) {
+    var solution = req.queryParam("id")
+        .map(Long::parseLong)
+        .map(solutionId -> metaSolver.getSolutionManager().getSolution(solutionId))
+        .map(Solution::toStringSolution)
+        .orElseThrow(() -> new ResponseStatusException(HttpStatus.NOT_FOUND,
+            "Could not find a solution for this problem with this solution id!"));
+
+    // yes, solution is of type `Solution<String>`.
+    // No idea why `toStringSolution` returns `SolutionHandle`
```

Comment:
Probably in a misguided effort to use the type with the fewest assumptions. Feel free to change that.